### PR TITLE
snes_ppu.cpp: Use device_palette_interface for Palette handling

### PR DESCRIPTION
**Pull request by cam900**
Fix CGRAM size, Direct color behavior
Reduce unnecessary values

### CODE REVIEW

**Comment by MooglyGuy** [comment]
I don't trust this change. Please provide a practical example of something it fixes, and test a large number of SNES games to ensure that it does not break anything. This seems like exactly the sort of change for change's sake that has a tendency to break things due to poor testing on your part.

**Comment by cam900** [comment]
This PR fixes direct color, and performance affects tiny. And it's also enable CGRAM viewer when you press F4.
ex: aerobiz title screen

File: src/devices/video/snes_ppu.cpp [hunk 1]
```diff
@@ -83,9 +83,9 @@
 #define SNES_CLIP_ALWAYS   3
 
 #define SNES_VRAM_SIZE        0x10000   /* 64kb of video ram */
-#define SNES_CGRAM_SIZE       0x202     /* 256 16-bit colours + 1 tacked on 16-bit colour for fixed colour */
+#define SNES_CGRAM_SIZE       0x200     /* 256 16-bit colours */
 #define SNES_OAM_SIZE         0x440     /* 1088 bytes of Object Attribute Memory */
-#define FIXED_COLOUR          256       /* Position in cgram for fixed colour */
+#define FIXED_COLOUR          0x900     /* Position in cgram for fixed colour */
```
Comment (cuavas) [inline src/devices/video/snes_ppu.cpp:88]: Is the comment still correct?  0x900 is too high to be an offset in a 256-word RAM.

**Review by cuavas — COMMENTED** [review]
It’s cleaner like this, and it will show the indirect palettes in the graphics viewer.

It needs proper testing across titles that use different colour modes.  What are a few games that exercise the different colour modes?  We need to see that CGRAM palette animation works correctly.  It should be fairly obvious if direct colour is broken.  It goes without saying that handling of the fixed colour should be tested.  Please ensure you’ve tested with representative cases and post results.

We’ll hold off merging this until at least after freeze.  There will be plenty of time before 0.227 to ensure it’s OK.

File: src/devices/video/snes_ppu.cpp (hunk 1 above)
Comment (cam900) [reply src/devices/video/snes_ppu.cpp:88]: In palette entry, 256-word CGRAM area is 0x00-0xff, Direct colour layout is 0x100-8ff (256 per group).
Fixed colour offset is the topmost entry of palette now.

**Comment by cam900** [comment]
aerobiz direct color is fixed.



Before

aerobiza
aerobizb
After

**Comment by cam900** [comment]
smw background color is same at both before and after applying PR.

File: src/devices/video/snes_ppu.h [hunk 2]
```diff
@@ -314,10 +314,14 @@ class snes_ppu_device :  public device_t,
...
 
 	// device_palette_interface overrides
-	virtual uint32_t palette_entries() const override { return 0x100 + (0x100 * 8) + 1; } // 256 word CG RAM data, 8 group of direct colours, Fixed color
+	// 256 word CG RAM data (0x000-0x0ff), 8 group of direct colours (0x100-0x8ff), Fixed color (0x900)
+	virtual uint32_t palette_entries() const override { return 0x100 + (0x100 * 8) + 1; }
 	virtual uint32_t palette_indirect_entries() const override { return 32 * 32 * 32; } // 15 bit BGR
 
 private:
+	const uint16_t DIRECT_COLOUR = 0x100; // Position in palette entry for direct colour
+	const uint16_t FIXED_COLOUR = 0x100 + (0x100 * 8); // Position in palette entry for fixed colour
```
Comment (cuavas) [inline src/devices/video/snes_ppu.h:323]: These should be `static constexpr` – you don’t want to create instance members for them.

**Comment by cuavas** [comment]
Is there a good test case for CGRAM palette animation (i.e. changing the colour of objects on the screen by writing to CGRAM)?

**Comment by cam900** [comment]
> Is there a good test case for CGRAM palette animation (i.e. changing the colour of objects on the screen by writing to CGRAM)?

Yoshi coin in smw, Working correctly and almost same.

**Pull request merged**